### ISSUE TITLE
Bug: The token type of `if` in the 262 `IdentifierName` production should be Identifier instead of Keyword

### ISSUE THREAD

**Issue by JLHwung** [typescript-eslint/typescript-eslint#11037]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

typescript-estree

### Playground Link

https://typescript-eslint.io/play#ts=5.6.2&showAST=es&fileType=.tsx&code=IYOglgZkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=true

### Repro Code

```TypeScript
a.if
```

### ESLint Config

```javascript

```

### tsconfig

```jsonc

```

### Expected Result

`.ast.tokens[2].type` should be `Identifier` (espree)

### Actual Result

`.ast.tokens[2].type` is `Keyword` (espree)

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/typescript-estree` | `8.29.1` |

**Comment by bradzacher** [comment]
Looking in ast-explorer espree emits `keyword` for both

https://astexplorer.net/#/gist/f6a5bdc607b61e885503ab484f571f63/8b439c436b610f9cc13ce84734a439273d8d9813

I've spoken to them about this before and essentially they said that they don't do contextual analysis of tokens to delineate between contextual keywords.

For example they also use the wrong type for `await`.

**Comment by JLHwung** [comment]
Thank you for triaging. Unfortunately the espree version on astexplorer.net is 9.6.1, which is a bit outdated. Could you check 10.3.0 on ast-explorer.dev?

**Comment by bradzacher** [comment]
Oh they must have changed it at some point.
See my other mention of tokens not being part of the ESTree spec so there's nothing governing what should be what.

I'm happy to change this -- but it might be a bit gnarly.

**Comment by ntdiary** [comment]
> Oh they must have changed it at some point.

The change was introduced in [this commit](https://github.com/acornjs/acorn/commit/38b7af471b245800bf733b4d34cfa9bbaca2dbae) to address a similar [issue](https://github.com/acornjs/acorn/issues/1232) (use keywords as property names, and treat them as `Identifier`).

I dug into some related issues and PRs in the acorn, espree, and  typescript-estree, and I feel there are two options to fix this issue:
1. Remove the possibly outdated logic below, I haven't found any valid case where we would need to convert a token that TS has already recognized as an `Identifier` back into a `Keyword`.
https://github.com/typescript-eslint/typescript-eslint/blob/76cc62cb37353711b64e3353bb62e2e0a688d629/packages/typescript-estree/src/node-utils.ts#L483-L501
2. Check `token.parent.kind`, if it's `PropertyAccessExpression`, we could just skip the conversion.
    ```ts
      if (isAtLeast50 && token.kind === SyntaxKind.Identifier && token.parent.kind !== SyntaxKind.PropertyAccessExpression) {
        keywordKind = ts.identifierToKeywordKind(token as ts.Identifier);
      } else if ('originalKeywordKind' in token && token.parent.kind !== SyntaxKind.PropertyAccessExpression) {
        // @ts-expect-error -- intentional fallback for older TS versions <=4.9
        keywordKind = token.originalKeywordKind;
      }
    ```
I personally lean towards option 1. If you're concerned it might cause regressions, option 2 is also fine.

BTW, there are 3 edge cases in espree that I believe should be fixed within their repository, they intended to support the `static/let/yield` keywords, just overlooked cases where they are used as property name:
```ts
a.static
a.yield
a.let
```
related code: https://github.com/eslint/js/blob/1d39fd2501722cd8ee32de07521ae8c18962f0a6/packages/espree/lib/token-translator.js#L109-L115

Please feel free to let me know your thoughts and then I could raise a PR here and create another issue in their repo later. :)

**Comment by ntdiary** [comment]
Hi, @JLHwung, my PR #11329 will treat all keywords used as property names as `Identifier`. However, when I added tests for this, I noticed that `@babel/eslint-parser` converts `static/let/yield` to `Keyword` as regardless of where they're used as property names (just like `eslint` does). Do you think we could fix this in the babel repo, or just leave it as is? Looking forward to hearing your thoughts. :)

https://github.com/babel/babel/blob/2ea3fc8f9b33a911840f17fbc407e7bfae2ed66f/eslint/babel-eslint-parser/src/convert/convertTokens.cts#L111-L119

**Comment by JLHwung** [comment]
@ntdiary Thanks for the heads-up. Yeah this was intentional, the `@babel/eslint-parser`'s job is to align with the `eslint` parser behavior, any difference is treated as bug, unless the production is from an ES proposal. If `eslint` will change that part, it would be great and Babel will surely follow. Maybe you can open an issue there.
